### PR TITLE
feat(storage): Add retry conformance test

### PR DESCRIPTION
**Pull request by bajajneha27**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [ ] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [ ] Update code documentation if necessary.

closes: b/228274210

### CODE REVIEW

File: google-cloud-storage/Rakefile [hunk 1]
```diff
@@ -26,6 +33,14 @@ namespace :test do
   end
 end
 
+#Conformance Tests
```
Comment (shivgautam) [inline google-cloud-storage/Rakefile:36]: Add space here

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+    bucket.reload!
+  end
+
+  def self.get_bucket_policy client, _preconditions, **resources
+    bucket_name = resources[:bucket].name
+    client.get_bucket_policy bucket_name
+  end
+
+  def self.insert_bucket client, _preconditions, **resources
+    bucket_name = "new_bucket" + Time.now.to_i.to_s
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:147]: How do we avoid conflicts / duplicate bucket names if multiple tests are running at the same time? This can get flaky.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+    if _preconditions
+      client.patch_bucket bucket.name, bucket_gapi, if_metageneration_match: metageneration
+    else
+      client.patch_bucket bucket.name, bucket_gapi
+    end
+  end
+
+  def self.bucket_acl_public client, _preconditions, **resources
+    bucket = resources[:bucket]
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:258]: Typically _varname is used when the variable is not used. Might want to change the variable name here.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+      bucket.acl.public! if_metageneration_match: bucket.metageneration
+    else
+      bucket.acl.public!
+    end
+  end
+
+  def self.bucket_acl_private client, _preconditions, **resources
+    bucket = resources[:bucket]
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:267]: Same here as above.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+
+  def self.project_service_account client, _preconditions, **resources
+    client.project_service_account
+  end
+
+  def self.patch_bucket client, _preconditions, **resources
+    bucket = resources[:bucket]
+    metageneration = resources[:bucket].metageneration
+    bucket_gapi = Google::Apis::StorageV1::Bucket.new storage_class: "COLDLINE"
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:249]: Here too.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+      bucket.acl.private! if_metageneration_match: bucket.metageneration
+    else
+      bucket.acl.private!
+    end
+  end
+
+  def self.default_acl_owner_full client, _preconditions, **resources
+    bucket = resources[:bucket]
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:276]: And here

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+    policy.etag = nil unless _preconditions
+    client.set_bucket_policy bucket_name, policy
+  end
+
+  def self.update_hmac_key client, _preconditions, **resources
+    access_id = resources[:hmac_key].access_id
+    etag = resources[:hmac_key].etag
+
+    hmac_key = Google::Apis::StorageV1::HmacKeyMetadata.new access_id: access_id, state: "INACTIVE"
+    hmac_key.etag = etag if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:301]: Here as well

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+  def self.create_notification client, _preconditions, **resources
+    bucket = resources[:bucket]
+    pubsub_topic_name = CONF_TEST_PUBSUB_TOPIC_NAME
+    bucket.create_notification pubsub_topic_name
+  end
+
+  def self.insert_object client, _preconditions, **resources
+    bucket = resources[:bucket]
+    file = StringIO.new CONF_TEST_FILE_CONTENT * 1024 * 1024 # 1MB
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:372]: Same here.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+      bucket.create_file file, CONF_TEST_FILE_PATH, if_generation_match: 0
+    else
+      bucket.create_file file, CONF_TEST_FILE_PATH
+    end
+  end
+
+  def self.get_object client, _preconditions, **resources
+    bucket = resources[:bucket]
+    object = resources[:object]
+    file = bucket.file object.name
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:382]: file variable is not needed

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+  def self.get_file client, _preconditions, **resources
+    bucket = resources[:bucket]
+    object = resources[:object]
+    client.get_file bucket.name, object.name
+  end
+
+  def self.download_file client, _preconditions, **resources
+    bucket = resources[:bucket]
+    object = resources[:object]
+    res = object.download
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:394]: res variable not needed

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+    role = "READER"
+    client.insert_file_acl bucket.name, object.name, entity, role
+  end
+
+  def self.compose_file client, _preconditions, **resources
+    bucket = resources[:bucket]
+    object = resources[:object]
+    object_2 = bucket.create_file StringIO.new(CONF_TEST_FILE_CONTENT), "my-test-file-2"
+    destination = "new-composite-object"
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:417]: underscored variable used.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+      bucket.compose [object.name, object_2.name], destination do |f|
+        f.content_type = "text/plain"
+      end
+    end
+  end
+
+  def self.delete_file client, _preconditions, **resources
+    bucket = resources[:bucket]
+    object = resources[:object]
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:431]: Same here. underscored variable

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+
+  def self.list_files client, _preconditions, **resources
+    bucket = resources[:bucket]
+    client.list_files bucket.name
+  end
+
+  def self.patch_file client, _preconditions, **resources
+    bucket = resources[:bucket]
+    object = resources[:object]
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:446]: Same here

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+      client.patch_file bucket.name, object.name, if_metageneration_match: object.metageneration
+    else
+      client.patch_file bucket.name, object.name
+    end
+  end
+
+  def self.rewrite_file client, _preconditions, **resources
+    bucket = resources[:bucket]
+    object = resources[:object]
+    if _preconditions
```
Comment (shivgautam) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:456]: And here

File: google-cloud-storage/Rakefile (hunk 1 above)
Comment (shivgautam) [inline google-cloud-storage/Rakefile:36]: Add space

File: google-cloud-storage/conformance/v1/proto/google/cloud/conformance/storage/v1/tests.proto [hunk 15]
```diff
@@ -60,21 +61,95 @@ message PolicyConditions {
...
   int32 expiration = 6;
   google.protobuf.Timestamp timestamp = 7;
+  /*
+    fields with strict equivalence which are added into
+    PolicyOutput.expectedDecodedPolicy to generate the
+    signature.
+
+    Expectations
+
+    E.1: Order them in lexigraphical order so it's the
```
[the remark this answers is not in the document]
Comment (dazuma) [reply google-cloud-storage/conformance/v1/proto/google/cloud/conformance/storage/v1/tests.proto:83]: I assume this proto file is just copied verbatim from storage's conformance repository, right? If so, we don't need to correct their grammar; I think it's better to ensure the files are identical.

File: google-cloud-storage/test/google/cloud/storage/retry/conformance_test.rb [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "helper"
+require "net/http"
+require_relative "../../../../../conformance/v1/proto/google/cloud/conformance/storage/v1/tests_pb"
+require_relative "./utils.rb"
+
+Google::Apis.logger.level = Logger::DEBUG
```
[the remark this answers is not in the document]
Comment (dazuma) [reply google-cloud-storage/test/google/cloud/storage/retry/conformance_test.rb:20]: Yes, I'm wary of this, especially since it's being done globally. Are there tests that assert against log output? Otherwise, I'd rather we didn't do this.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb (hunk 2 above)
Comment (dazuma) [reply google-cloud-storage/test/google/cloud/storage/retry/utils.rb:147]: Agreed. It's best to append some random digits to prevent this.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb (hunk 3 above)
Comment (dazuma) [reply google-cloud-storage/test/google/cloud/storage/retry/utils.rb:258]: +1

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+
+  def self.create_notification client, _preconditions, **resources
+    bucket = resources[:bucket]
+    pubsub_topic_name = CONF_TEST_PUBSUB_TOPIC_NAME
+    bucket.create_notification pubsub_topic_name
+  end
+
+  def self.insert_object client, _preconditions, **resources
+    bucket = resources[:bucket]
+    file = StringIO.new CONF_TEST_FILE_CONTENT * 1024 * 1024 # 1MB
```
Comment (dazuma) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:371]: Note this is actually `1024 * 1024 * CONF_TEST_FILE_CONTENT.size` which is 12MB rather than 1MB. Just making sure that's what you intend for this object.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb (hunk 2 above)
Comment (shivgautam) [reply google-cloud-storage/test/google/cloud/storage/retry/utils.rb:147]: Yes. random upto 10000 should be a safer bet

File: google-cloud-storage/test/google/cloud/storage/retry/conformance_test.rb [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+  def self.run_test_case test_name, scenario, instructions, method, lib_func
+    define_method("test_#{test_name}") do
+      expect_success = scenario.expectSuccess
+
+      response = create_retry_test method.name, instructions
+      test_id = JSON.parse(response.body)["id"]
+
+      success_result = true
+      begin
+        run_retry_test test_id, lib_func, scenario["preconditionProvided"], method.resources
```
[the remark this answers is not in the document]
Comment (bajajneha27) [reply google-cloud-storage/test/google/cloud/storage/retry/conformance_test.rb]: No specific reason. Since it was a JSON key, I used the string

File: google-cloud-storage/test/google/cloud/storage/retry/conformance_test.rb [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+##################################################################################################
+### Helper Methods for Testbench Retry Test API ##################################################
+##################################################################################################
+
+##
+# The Retry Test API in the testbench is used to run the retry conformance tests. It offers a 
+# mechanism to describe more complex
+# retry scenarios while sending a single, constant header through all the HTTP requests from a
+# test program. The Retry Test API can be accessed by adding the path "/retry-test" to the host. # See also: https://github.com/googleapis/storage-testbench
```
[the remark this answers is not in the document]
Comment (bajajneha27) [reply google-cloud-storage/test/google/cloud/storage/retry/conformance_test.rb]: Done

File: google-cloud-storage/test/helper.rb [hunk 20]
```diff
@@ -42,6 +42,10 @@ class MockStorage < Minitest::Spec
   let(:project) { "test" }
   let(:credentials) { OpenStruct.new(client: OpenStruct.new(updater_proc: Proc.new {})) }
   let(:storage) { Google::Cloud::Storage::Project.new(Google::Cloud::Storage::Service.new(project, credentials)) }
+  let(:pubsub_topic_name) { "my-topic-name" }
```
[the remark this answers is not in the document]
Comment (bajajneha27) [reply google-cloud-storage/test/helper.rb:45]: `pubsub_topic_name` is used to create notifications in storage. It's being used in [conformance_test.rb](https://github.com/googleapis/google-cloud-ruby/pull/18230/files#diff-1d5ff2a8c2cf2d50b8eeeae343b0c693b962b626409dd3149b37e2cbca377893R36)
and same goes with all other new variables added.
This helper file is common to all other tests, other variables are also used in other tests.

File: google-cloud-storage/Rakefile (hunk 1 above)
Comment (bajajneha27) [reply google-cloud-storage/Rakefile]: Done

File: google-cloud-storage/test/helper.rb (hunk 20 above)
[the remark this answers is not in the document]
Comment (shivgautam) [reply google-cloud-storage/test/helper.rb:45]: Ok

**Review by shivgautam — APPROVED** [review]
LGTM

**Review by dazuma — APPROVED** [review]
LGTM

File: google-cloud-storage/conformance/README.md [hunk 21]
```diff
@@ -15,3 +15,5 @@ The output directory is the same is the input directory, so after successful com
 ## Running the tests
 
 The conformance tests are included in the ordinary unit test suite. In the `google-cloud-storage` directory, run `rake test`.
+
```
Comment (tritone) [inline google-cloud-storage/conformance/README.md:18]: Does the emulator need to be launched separately in order to run the tests? If so worth noting how to do that.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 22]
```diff
@@ -0,0 +1,464 @@
...
+
+  # Method invocation mapping is a map whose keys are a string describing a standard
+  # API call (e.g. storage.objects.get) and values are a list of functions which
+  # wrap library methods that implement these calls. There may be multiple values
+  # because multiple library methods may use the same call (e.g. get could be a
+  # read or just a metadata get).
+
+  def self.get
+  {
+    "storage.bucket_acl.list" => [:list_bucket_acls],
```
Comment (tritone) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:37]: Nit: could you sort these by resource / alphabetically? Otherwise it's a bit difficult to see all of what's implemented.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb [hunk 23]
```diff
@@ -0,0 +1,464 @@
...
+    "storage.buckets.patch" => [
+      :patch_bucket,
+      :bucket_acl_public,
+      :bucket_acl_private,
+      :default_acl_owner_full,
+    ],
+    "storage.buckets.setIamPolicy" => [:set_bucket_policy],
+    "storage.hmacKey.update" => [:update_hmac_key],
+    "storage.resumable.upload" => [
+      :blob_upload_from_string,
```
Comment (tritone) [inline google-cloud-storage/test/google/cloud/storage/retry/utils.rb:91]: Also, I notice that I think these still need to be implemented, correct? Fine to do it in a separate PR.

File: google-cloud-storage/conformance/README.md (hunk 21 above)
Comment (bajajneha27) [reply google-cloud-storage/conformance/README.md:18]: Yes, it does. Updated README to include that. Thanks.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb (hunk 22 above)
Comment (bajajneha27) [reply google-cloud-storage/test/google/cloud/storage/retry/utils.rb:37]: Makes sense. Updated it alphabetically.

File: google-cloud-storage/test/google/cloud/storage/retry/utils.rb (hunk 23 above)
Comment (bajajneha27) [reply google-cloud-storage/test/google/cloud/storage/retry/utils.rb:91]: I've implemented them as part of `storage.objects.insert` . If we need to implement more varieties of it, can be done in separate PR.
